examples/example_customdns_test.go

```diff
@@ -0,0 +1,165 @@
+// Copyright (C) MongoDB, Inc. 2022-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package examples
+
+import (
+	"context"
+	"net"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/miekg/dns"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+func resolve(ctx context.Context, cache *dnsCache, in *dns.Conn, out *dns.Conn) {
+	for ctx.Err() == nil {
+		q, err := in.ReadMsg()
+		if err != nil {
+			// TODO: Handle error.
+			continue
+		}
+		if len(q.Question) != 1 {
+			// Multiple questions in a single query is not actually used in real life.
+			continue
+		}
+
+		a, err := func() (*dns.Msg, error) {
+			cache.lock.Lock()
+			defer cache.lock.Unlock()
+
+			now := time.Now()
+			if rr, ok := cache.records[q.Question[0]]; ok && rr.exp.After(now) {
+				a := new(dns.Msg)
+				a.SetReply(q)
+				a.Compress = false
+				a.Answer = append(a.Answer, rr.record)
+				return a, nil
+			}
+
+			err := out.WriteMsg(q)
+			if err != nil {
+				return nil, err
+			}
+
+			m, err := out.ReadMsg()
+			if err != nil {
+				return nil, err
+			}
+
+			l := len(m.Answer)
+			for i, q := range m.Question {
+				if i >= l {
+					break
```

Comment:
Is it an error if less answers than questions are received? If so, I suggest adding the comment: `TODO: Handle error.` and a `log.Fatalf`.

Reply:
It may not be an error (https://superuser.com/questions/523917/dns-queries-returning-no-answer-section). The DNS protocol itself is complex enough to have its package, and which is also a reason for not including the cache in the driver if the use case requires a proper TTL (mostly meaning a fully functional and compatible client implement).